packages/cojson/src/sync.ts

```diff
@@ -285,27 +291,85 @@ export class SyncManager {
         logger.error("Error sending new content piece, retrying", { err: e });
         peer.optimisticKnownStates.dispatch({
           type: "SET",
-          id,
-          value: optimisticKnownStateBefore ?? emptyKnownState(id),
+          id: coValue.id,
+          value: optimisticKnownStateBefore ?? emptyKnownState(coValue.id),
         });
-        return this.sendNewContentIncludingDependencies(id, peer);
+        return this.sendNewContent(coValue, peer);
       });
 
       peer.optimisticKnownStates.dispatch({
         type: "COMBINE_WITH",
-        id,
+        id: coValue.id,
         value: coValue.knownState(),
       });
     }
   }
 
-  addPeer(peer: Peer) {
+  async startPeerReconciliation(peer: PeerState) {
+    const coValuesOrderedByDependency: CoValueCore[] = [];
+
+    const requested = new Set<string>();
+
+    const navigateDeps = (coValue: CoValueCore) => {
+      if (requested.has(coValue.id)) {
+        return;
+      }
+
+      requested.add(coValue.id);
+
+      for (const id of coValue.getDependedOnCoValues()) {
+        const entry = this.local.coValuesStore.get(id);
+
+        if (entry.state.type === "available") {
+          navigateDeps(entry.state.coValue);
+        }
+      }
+
+      coValuesOrderedByDependency.push(coValue);
+    };
+
+    for (const entry of this.local.coValuesStore.getValues()) {
+      switch (entry.state.type) {
+        case "unavailable":
+          if (!peer.toldKnownState.has(entry.id)) {
+            await entry.loadFromPeers([peer]).catch((e: unknown) => {
+              logger.error("Error sending load", { err: e });
+            });
+          }
+          break;
+        case "available":
+          const coValue = entry.state.coValue;
+
+          navigateDeps(coValue);
+          break;
+      }
+
+      if (!peer.optimisticKnownStates.has(entry.id)) {
+        peer.optimisticKnownStates.dispatch({
+          type: "SET_AS_EMPTY",
+          id: entry.id,
+        });
+      }
+    }
+
+    for (const coValue of coValuesOrderedByDependency) {
+      await this.trySendToPeer(peer, {
+        action: "load",
+        ...coValue.knownState(),
+      }).catch((e: unknown) => {
+        logger.error("Error sending load", { err: e });
+      });
+    }
+  }
+
+  async addPeer(peer: Peer) {
     const prevPeer = this.peers[peer.id];
     const peerState = new PeerState(peer, prevPeer?.knownStates);
     this.peers[peer.id] = peerState;
 
     if (prevPeer && !prevPeer.closed) {
       prevPeer.gracefulShutdown();
+      await prevPeer.processMessages?.catch((e) => {});
```

Comment:
We need to wait for the previous peer to process all the incoming messages before starting to process the new  peer messages.

This is probably why, on frequent reconnections, the sync server starts to have troubles.

Reply:
Quick fix, doesn't handle more than one peers in line to be added.

Will improve in the final version of the PR